src/pip/_internal/index.py

```diff
@@ -296,6 +298,35 @@ def _check_link_requires_python(
     return True
 
 
+def get_best_from_sorted(
+    candidates,          # type: List[InstallationCandidate]
+    allow_yanked=False,  # type: bool
+):
+    # type: (...) -> Optional[InstallationCandidate]
+    """
+    Return the best candidate, using whether the candidates are yanked.
+
+    :param candidates: Installation candidates sorted from most preferred
+        to least preferred.
+    :param allow_yanked: Whether to permit returning a yanked candidate
+        in the sense of PEP 592. If true, a yanked candidate will be
+        returned only if all candidates have been yanked.
+    """
+    if not candidates:
+        return None
+
+    for candidate in candidates:
+        if not candidate.location.is_yanked:
```

Comment:
This is what triggered that long comment requesting to model this differently -- this is doing filtering in `CandidateEvalutator`.